core/src/main/java/com/linecorp/armeria/server/Route.java

```diff
@@ -209,4 +213,18 @@ default RoutingResult apply(RoutingContext routingCtx) {
      * </ul>
      */
     Route withPrefix(String prefix);
+
+    /**
+     * A list of predicates used to determine whether a {@link Route} will accept a request
+     * based on its HTTP parameters.
+     */
+    @UnstableApi
+    List<? extends Predicate<QueryParams>> paramPredicates();
```

Comment:
I think the main problem is that `DefaultRoutingContext.equals()` excludes query params or headers from comparison. It makes sense since the cardinality could be huge and the cache hit ratio could be low.

Building `dynamicPredicateRoutes` as a separate list itself sounds also good. But `Route` should know if the route could be cacheable. So what do you think of adding `isCacheable()` property to `Route` instead?

Additionally, I am planning to change the type of `Predicate<QueryParams>` to a data structure instead of a function so that we can extract the properties to the service specification. There is also a need to minimize breaking change for type changes.

Reply:
I understood your comment to just expose an `isCacheable` method instead of each `*predicates`.

Route.java
```java 
bool isCacheable() {
  return route.paramPredicates().isEmpty() &&
    route.headerPredicates().isEmpty() &&
    route.excludedRoutes().isEmpty());
}
```

Sure, I'm fine with either way